### PR TITLE
[Feature] New RBAC privilege framework: support user with host

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Introduce a new privilege framework.
To avoid huge PR which is hard to review and easy to conflict, I'm splitting into several small PR, each implementing an individual aspect of the new privilege framework. This is the eleventh one. For previous PRs see 
https://github.com/StarRocks/starrocks/pull/11447
https://github.com/StarRocks/starrocks/pull/11528
https://github.com/StarRocks/starrocks/pull/11593
https://github.com/StarRocks/starrocks/pull/11732
https://github.com/StarRocks/starrocks/pull/11777
https://github.com/StarRocks/starrocks/pull/11847
https://github.com/StarRocks/starrocks/pull/11909
https://github.com/StarRocks/starrocks/pull/12035
https://github.com/StarRocks/starrocks/pull/12128
https://github.com/StarRocks/starrocks/pull/12172

This PR support user with hostname such as `user@['host01']` and implements asynchronous resolving hostname to ip.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 53 / 57 (92.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/DomainResolver.java|10|12|83.33%|[75, 95]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|40|42|95.24%|[51, 147]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/UserAuthenticationInfo.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 56 / 58 (96.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/DomainResolver.java|11|13|84.62%|[75, 95]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|42|42|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/UserAuthenticationInfo.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 55 / 56 (98.21%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|3|66.67%|[1318]|
|:large_blue_circle:|com/starrocks/catalog/DomainResolver.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/UserAuthenticationInfo.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/authentication/AuthenticationManager.java [hunk 1]
```diff
@@ -96,16 +117,52 @@ public String getDefaultPlugin() {
...
+    }
+
+    private boolean match(String remoteUser, String remoteHost, boolean isDomain, UserAuthenticationInfo info) {
+        // quickly filter unmatched entries by user name
+        if (!info.matchUser(remoteUser)) {
+            return false;
+        }
+        if (isDomain) {
+            // check for resolved ips
+            Set<String> ipSet = hostnameToIpSet.get(info.getOrigHost());
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/authentication/AuthenticationManager.java:145]: do we need hostnameToIpLock.readLock() here?

**Review by nshangyiming — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/authentication/AuthenticationManager.java (hunk 1 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/authentication/AuthenticationManager.java]: My mistake. Done.

**Review by nshangyiming — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 2]
```diff
@@ -1113,10 +1113,8 @@ private void startNonLeaderDaemonThreads() {
             // metastoreEventsProcessor.init();
             // metastoreEventsProcessor.start();
         }
-        if (!usingNewPrivilege) {
-            // domain resolver
-            domainResolver.start();
-        }
+        // domain resolver
+        domainResolver.start();
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1117]: domainResolver only inited when usingNewPrivilege is true. should check nullable?

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:1117]: domainResolver will initialize no matter whether usingNewPrivilege is true or false. I have two separate constructors for each case.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 58 (98.28%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|3|66.67%|[1318]|
|:large_blue_circle:|com/starrocks/catalog/DomainResolver.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationManager.java|40|40|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/UserAuthenticationInfo.java|2|2|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Pull request merged**